### ISSUE TITLE
Something is wrong of my offset flow during training

### ISSUE THREAD

**Issue by zrfreya** [JiahuiYu/generative_inpainting#221]
I am training the DeepFillv2 with my own data. During the training, something is wrong in the visualized offset flow of attention map. There are color pixels in the sky region where is not in the mask. I assume that there should be white in all non-mask regions. Here are my results.
![single image](https://user-images.githubusercontent.com/17635831/53889576-9ce88400-4061-11e9-8fd7-abb1c9c0e01b.png)

In my implementation, masks are different in one batch. I change the attention code as following and take mi in m_groups as mm in the original code:
`   int_ms = mask.get_shape().as_list()
    m = tf.extract_image_patches(
        mask, [1,ksize,ksize,1], [1,stride,stride,1], [1,1,1,1], padding='SAME')
    m = tf.reshape(m, [int_ms[0], -1, ksize, ksize, int_ms[3]])
    m = tf.transpose(m, [0, 2, 3, 4, 1])  # transpose to b*k*k*c*hw

    mm = tf.cast(tf.equal(tf.reduce_mean(m, axis=[1,2,3], keep_dims=True), 0.), tf.float32)
    m_groups = tf.split(mm, int_ms[0], axis=0)`

**Comment by JiahuiYu** [comment]
That’s strange, since we directly use mask to do contextual attention, instead of using condition that if pixel values are zero. I would suggest to debug by printing out exact values in contextual attention layer step by step using tf eager mode.

**Comment by zrfreya** [comment]
I test the original contextual attention layer with the following imageA imageB and mask. The last image is the output.
![1_170908_082216890_camera_5_mask](https://user-images.githubusercontent.com/17635831/53969581-0638c780-4134-11e9-9162-00d37f1b2ffa.png)
![1](https://user-images.githubusercontent.com/17635831/53969597-0c2ea880-4134-11e9-8823-ac76acdc9747.jpg)
![4_170908_075043345_camera_5_mask](https://user-images.githubusercontent.com/17635831/53969860-8101e280-4134-11e9-8bca-e331f0f96c6c.png)

![test_attention](https://user-images.githubusercontent.com/17635831/53969896-92e38580-4134-11e9-8ba1-4ab224083f2b.png)
 The result is very noisy and I can not see the influence of the mask. Is this normal?  @JiahuiYu

**Comment by JiahuiYu** [comment]
Hi it may because two images do not have similar patches. It should be fine since we have a decoder and another parallel branch in our Inpainting network.

**Comment by zrfreya** [comment]
> Hi it may because two images do not have similar patches. It should be fine since we have a decoder and another parallel branch in our Inpainting network.

Thanks for your reply. You are right. I use the same image for imageA and imageB, then the result is right. I tested the original contextual attention layer and my slightly-modified code with the same image. And I got same results. It means that my contextual attention layer is not wrong. But I still got wrong attention map in some images like the color dots in the sky region. Besides, the inpainted region (at the center of the image) is very noisy. 
![image](https://user-images.githubusercontent.com/17635831/54012225-5fded780-41b0-11e9-8d96-4ff927478e68.png)

And here are my loss curves:
![image](https://user-images.githubusercontent.com/17635831/54012399-e4c9f100-41b0-11e9-80aa-de94828d4723.png)

@JiahuiYu Do you have any clue about this situation? Will total_variation loss help to reduce the noise?  Thanks for your patient help.

**Comment by JiahuiYu** [comment]
1. Unfortunately I don’t have any idea why your attention visualization has this bug. If you finally find out the reason, I will be grateful if you can post the reason here. Thanks.

2. Yes that might helps a little. But before adding tv loss, I would suggest to tune loss balance hyper-parameter between GAN loss and reconstruction loss first. I would say in most cases, two losses are enough.

**Comment by zrfreya** [comment]
@JiahuiYu  Thank you for your advice. I will try your suggestion and keep on digging the attention visualization problem.

**Comment by neouyghur** [comment]
@zrfreya Hi I am also training the deepfill2. My curves are similar to yours. I feel except L1 loss other losses are fluctuating.  If I cancel out L1 loss generator and discriminator losses are changing. I would like to learn about your training. Thanks.
